### PR TITLE
Feature/cif write

### LINKED ISSUE #6: Write out a real .cif file with partial charge values.
I have a PR for this. 

The output file will have the same name but with the following extension: .fw2.cif

The one feature missing is a "chargefw2" category with all the meta data needed to reproduce the calculation.

### PR DESCRIPTION
**Pull request by danny305**
This PR should close #6. 

Here is my implementation of outputting a cif file. Let me know if there are any refactoring you want me to make so it fits your repo better. 

Also this was built on top of my [docker PR](https://github.com/krab1k/ChargeFW2/pull/5) so it works in Docker already. 

Like in the docker branch, the repo being cloned in the docker file should be refactored your repo's master branch. Right now it points to my repo and my feature/cif-write branch.

### CODE REVIEW

**Comment by krab1k** [comment]
Seems to be working at the first glance, nice. Please rebase it onto the current master. Will look more at the code, now just a few first stylish things.

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -32,6 +32,7 @@ add_dependencies(chargefw2 ${METHODS})
 
 target_link_libraries(chargefw2 structures ${COMMON_LIBS} formats utility lhs Boost::program_options dl Boost::filesystem fmt::fmt)
 
+
```
Comment (krab1k) [inline src/CMakeLists.txt:35]: That's unnecessary.

File: src/formats/cif.cpp [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+                        return true;
+                    }
+                }
+            }
+        }
+        return false;
+    }
+
+    const int _model;
+    const std::string _chain, _residue, _res_num, _atom;
```
Comment (krab1k) [inline src/formats/cif.cpp:106]: Please move attributes (and private things) to the top of the class. Also one attribute per line.

File: src/formats/cif.cpp [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+#include <string>
+#include <vector>
+#include <stdexcept>
+#include <fstream>
+#include <filesystem>
+
+#include <gemmi/cif.hpp>
+#include <gemmi/mmcif.hpp>
+
+#define GEMMI_WRITE_IMPLEMENTATION
```
Comment (krab1k) [inline src/formats/cif.cpp:15]: This is not available in the old gemmi library (0.3.6). Update requirements in README to some current version like 0.4.7.

File: src/formats/cif.cpp [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+
+class MCRA {
+public:
+    MCRA(const int model,
+         const std::string &chain,
+         const std::string &res_num,
+         const std::string &residue,
+         const std::string &atom)
+        : _model(model), _chain(chain), _res_num(res_num), _residue(residue), _atom(atom)
```
Comment (krab1k) [inline src/formats/cif.cpp:51]: Won't compile with GCC:
error: ‘MCRA::_res_num’ will be initialized after [-Werror=reorder]

File: src/formats/cif.cpp [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+            p_charge[row_num]  = std::to_string(chg[i]);
+            vdw_radii[row_num] = std::to_string(atom.element().vdw_radius());
+        }
+
+        write_cif_block(out_stream, table, p_charge, vdw_radii);
+    }
+    catch (std::out_of_range &) {
+        /* Do nothing */
+    }
+}
```
Comment (krab1k) [inline src/formats/cif.cpp:205]: Add a newline.

File: src/formats/cif.h [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+#include "../structures/molecule_set.h"
+#include "../charges.h"
+
+
+class CIF : public Writer {
+public:
+    void save_charges(const MoleculeSet &ms, const Charges &charges, const std::string &filename) override;
+private:
+    void write_cif_block(std::ostream &out, gemmi::cif::Table &table, std::vector<std::string> &p_charge, std::vector<std::string> &vdw_radii);
+};
```
Comment (krab1k) [inline src/formats/cif.h:21]: Add a newline.

File: src/formats/cif.cpp [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+    newCols[new_tag_size - 1] = std::move(vdw_radii);
+
+    // TODO get this info into the column name
+    fs::path params{config::par_file};
+    std::cout << "Method name: " << config::method_name << " Parameter Name: " << params.stem() << std::endl;
+
+    std::vector<std::string> new_tags{
+        "_atom_site.fw2_charge",
+        "_atom_site.fw2_vdw_radius"};
+    for (auto tag : new_tags)
```
Comment (krab1k) [inline src/formats/cif.cpp:141]: Add const & just to shut the clang-tidy.

File: src/formats/cif.cpp [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+    MCRA(const int model,
+         const std::string &chain,
+         const std::string &res_num,
+         const std::string &residue,
+         const std::string &atom)
+        : _model(model), _chain(chain), _res_num(res_num), _residue(residue), _atom(atom)
+        {}
+
+    int find_row(gemmi::cif::Table &table, size_t start_idx = 0) const {
+        size_t idx = start_idx, total_rows = table.length();
```
Comment (krab1k) [inline src/formats/cif.cpp:55]: One def per line.

Comment (krab1k) [reply src/formats/cif.cpp:55]: size_t idx produces quite a few warnings when converting to int.

File: src/formats/cif.cpp [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+        auto iCol = table.bloc.find_loop(loop.tags[i]);
+        std::copy(iCol.begin(), iCol.end(), newCols[i].begin());
+    }
+
+    newCols[new_tag_size - 2] = std::move(p_charge);
+    newCols[new_tag_size - 1] = std::move(vdw_radii);
+
+    // TODO get this info into the column name
+    fs::path params{config::par_file};
+    std::cout << "Method name: " << config::method_name << " Parameter Name: " << params.stem() << std::endl;
```
Comment (krab1k) [inline src/formats/cif.cpp:136]: This is already a part of the output. Also when no parameters are specified, parameters are selected automatically but params.stem() returns nothing.

File: src/CMakeLists.txt (hunk 1 above)
Comment (danny305) [reply src/CMakeLists.txt:35]: I am assuming you are talking about line 35?

File: src/formats/cif.cpp (hunk 2 above)
Comment (danny305) [reply src/formats/cif.cpp:106]: sure thing.

File: src/formats/cif.cpp (hunk 3 above)
Comment (danny305) [reply src/formats/cif.cpp:15]: Sure.

**Comment by danny305** [comment]
Let me know what else you need me to do. Take your time reviewing the code.

File: src/formats/cif.cpp [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+#include "chargefw2.h"
+#include "cif.h"
+#include "../structures/molecule_set.h"
+#include "../charges.h"
+#include "../config.h"
+
+
+namespace fs = std::filesystem;
+
+static const std::vector<std::string> atom_site_columns{
```
Comment (krab1k) [inline src/formats/cif.cpp:28]: Do we expect these columns are required/always present?

File: src/formats/cif.cpp [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+                          std::vector<std::string> &vdw_radii) {
+
+    auto &loop = *table.get_loop();
+
+    unsigned long orig_tag_size = loop.tags.size();
+    unsigned long new_tag_size = orig_tag_size + 2;
+
+    // Creates a new table full of empty strings with the correct number of dimensions
+    // Outside vector size is the # of columns, inside vector size is the # of rows.
+    std::vector<std::vector<std::string>> newCols(new_tag_size, {loop.length(), {"Empty"}});
```
Comment (krab1k) [inline src/formats/cif.cpp:126]: Please use camel_case for variables (newCols, iCol) to be consistent with the rest.

File: src/formats/cif.cpp [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+        auto iCol = table.bloc.find_loop(loop.tags[i]);
+        std::copy(iCol.begin(), iCol.end(), newCols[i].begin());
+    }
+
+    newCols[new_tag_size - 2] = std::move(p_charge);
+    newCols[new_tag_size - 1] = std::move(vdw_radii);
+
+    std::vector<std::string> new_tags{
+        "_atom_site.fw2_charge",
+        "_atom_site.fw2_vdw_radius"};
```
Comment (krab1k) [inline src/formats/cif.cpp:140]: Do we need the radius? I know PQR writer has that, but the goal was just to comply with "definitions" here:
https://www.ics.uci.edu/~dock/manuals/apbs/html/user-guide/a2566.html
or here: https://docs.mdanalysis.org/1.0.1/documentation_pages/coordinates/PQR.html

Do you need that for some application in the cif?

File: src/formats/cif.cpp [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+    }
+
+    auto p_charge  = std::vector<std::string>{table.length(), "?"};
+    auto vdw_radii = std::vector<std::string>{table.length(), "?"};
+
+    const auto &molecule = ms.molecules()[0];
+
+
+    // ChargeFW2 is hardcoded to only read first model.
+    const int model{1};
```
Comment (krab1k) [inline src/formats/cif.cpp:173]: Please use initialization with = for these simple variables.

File: src/formats/cif.cpp [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+                    }
+                }
+            }
+        }
+        return false;
+    }
+};
+
+
+void CIF::write_cif_block(std::ostream &out,
```
Comment (krab1k) [inline src/formats/cif.cpp:114]: I believe that the output CIF file should be again directly usable as an input for ChargeFW2 - right now the new tags are added again, resulting in GEMMI not reading the file correctly:

```
doc = gemmi.cif.read_file('/tmp/1rud.fw2.fw2.cif') 
RuntimeError: /tmp/1rud.fw2.fw2.cif:1592 in data_1RUD: duplicate tag _atom_site.fw2_charge
```

Maybe overwrite the existing values?

File: src/formats/cif.cpp [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+         const std::string &chain,
+         const std::string &res_num,
+         const std::string &residue,
+         const std::string &atom)
+        : _model(model), _chain(chain), _res_num(res_num), _residue(residue), _atom(atom)
+        {}
+
+    int find_row(gemmi::cif::Table &table, int start_idx = 0) const {
+        int idx = start_idx; 
+        int total_rows = table.length();
```
Comment (krab1k) [inline src/formats/cif.cpp:62]: static_cast<int> to silence the warning.

File: src/formats/cif.cpp [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+    }
+
+    friend std::ostream &operator<<(std::ostream &os, const MCRA &mcra) {
+        os << "Atom(" << mcra._model << " " << mcra._chain << " " << mcra._res_num << " [" << mcra._residue << "] " << mcra._atom << ")";
+        return os;
+    }
+
+private:
+    bool is_row(const gemmi::cif::Table::Row &row) const {
+        int model = std::stoul(row[11]);
```
Comment (krab1k) [inline src/formats/cif.cpp:94]: static_cast<int> to silence the warning.

File: src/CMakeLists.txt (hunk 1 above)
Comment (krab1k) [reply src/CMakeLists.txt]: Yep.

**Comment by krab1k** [comment]
OK, I looked a bit more into the PR. Other than the duplicated tag issue and some minor things, the rest seems fine to me.

In the end, please squash all the commits related to this CIF write feature into a single commit as I'd like to have a clean git history if it is possible. Thanks!

(Not sure why GitHub shows already merged commits in PR, that's a bit confusing...)

File: src/formats/cif.cpp (hunk 10 above)
Comment (danny305) [reply src/formats/cif.cpp:28]: if the cif file is from the PDB then these columns will always be present. These columns adhere from the mmCIF requirements. Maybe the solution would be to only write out the cif file if the input is a cif and it is a protein. This should be controlled in main.cpp.

File: src/formats/cif.cpp (hunk 12 above)
Comment (danny305) [reply src/formats/cif.cpp:140]: I personally do not use the radius. This is was inherited from my implementation of this feature in[ freesasa](https://github.com/mittinatten/freesasa/blob/master/src/cif.cc). 

I am unsure if anyone currently needs it for a cif application, its more for being consistent and enabling people to move to cif in the future where they currently use pqr. 

Let me know if you want to remove it. In my opinion, I think its best to leave it for other people whom might need it.

File: src/formats/cif.cpp (hunk 13 above)
Comment (danny305) [reply src/formats/cif.cpp:173]: sure.

File: src/formats/cif.cpp (hunk 14 above)
Comment (danny305) [reply src/formats/cif.cpp:114]: This is a nice catch! I never thought about this and is probably a bug also in freesasa.

**Comment by danny305** [comment]
I addressed everything. I am unsure how to squash the commits on my end.

**Comment by danny305** [comment]
Yo Tomas, I added all the configuration data used to calculate the charges to the out cif file. It should be at the very bottom of the out .fw2.cif file produced. Let me know if I am missing any data you think should be added.

**Comment by krab1k** [comment]
Thanks! I merged the PR and made a few adjustments myself to speed things up. Most notably - the name of the parameters that were used is correctly written even if the user did not specify them. I also remove ref_chg_file since it has no relevance for the charge calculation (it's used only when comparing two sets of charges).

Anyway, thanks again for the contribution!

**Pull request merged**